### PR TITLE
Add my encoding debugging rig

### PR DESCRIPTION
**Pull request by saethlin**
Let's see if this has significant overhead. I've never checked.

r? @ghost

### CODE REVIEW

**Comment by workingjubilee** [comment]
...huh, can we figure out what is going on with the perf on bitmaps?

**Comment by saethlin** [comment]
The perf run above is informative because it says there's no significant perf issues with this when not in use. So this general design is a good candidate for actually merging.

bitmaps incr-unchanged is a stress-test for incr comp dep grah encoding. The majority of encoder traffic goes through `write_with` or `write_all`, and all those calls now need to and/or their bounds check with a bool. There's no impact on cycles, so I don't think the icount regression is that important. We are probably getting a free throughput increase from speculation/branch prediction because that bool is always false.

That being said, this code has long bothered me:
https://github.com/rust-lang/rust/blob/084ce5bdb5f7dc1c725f6770a8de281165ba3b0a/compiler/rustc_query_system/src/dep_graph/serialized.rs#L402-L407

So before I just ask people to accept the red perf result, I'm going to see if I can batch those encodings.

There's one issue with the implementation I need to fix too. This path can cause the desired offset to be jumped past: https://github.com/rust-lang/rust/blob/084ce5bdb5f7dc1c725f6770a8de281165ba3b0a/compiler/rustc_serialize/src/opaque.rs#L108

File: compiler/rustc_serialize/src/opaque.rs [hunk 1]
```diff
@@ -75,6 +82,19 @@ impl FileEncoder {
         {
             self.finished = false;
         }
+
+        if let Some(panic_offset) = self.panic_at_offset {
+            // If we are within the buffer size of the offset we're to panic at, we need to start
+            // flushing every write so that we don't panic late.
+            if self.flushed - panic_offset <= BUF_SIZE {
```
Comment (JakobDegen) [inline compiler/rustc_serialize/src/opaque.rs:89]: I think the subtraction is backwards

File: compiler/rustc_serialize/src/opaque.rs [hunk 2]
```diff
@@ -57,6 +62,8 @@ impl FileEncoder {
             res: Ok(()),
             #[cfg(debug_assertions)]
             finished: false,
+            panic_at_offset,
+            flush_every_write: false,
```
Comment (JakobDegen) [inline compiler/rustc_serialize/src/opaque.rs:66]:
```suggestion
panic_at_offset.is_some_and(|v| v <= BUF_SIZE)
```

**Comment by JakobDegen** [comment]
In addition to what you mentioned, the `if` branch in `write_all_cold_path` also needs a
```
            if self.flush_every_write {
                self.flush();
            }
```
I think

**Comment by JakobDegen** [comment]
Actually, the `write_with` case is kind of backwards. `flush_every_write` will force every write to be flushed before something is appended to the buffer, but then once it's appended to the buffer there's no subsequent `flush`. This means we'll actually panick on the next write after the one that we care about

**Comment by saethlin** [comment]
Yeah, a test or two would probably help.

**Comment by saethlin** [comment]
Going to see if batching writes for the dep graph fixes the perf change, noodle on the issue Jake pointed out.
@bors try @rust-timer queue

**Comment by saethlin** [comment]
Lol I made `bitmaps` faster and almost everything else slower.

**Comment by saethlin** [comment]
Some quick `counts -w` work indicates that `bitmaps` is the only benchmark where large edge lists dominate. Of course.

**Comment by saethlin** [comment]
Ah! The regression is driven by Span encoding. I think that needs some batching too...

**Comment by saethlin** [comment]
This has rotted quite a bit, and not really proved its value. Bummer.

**Pull request closed without merging**